File: python/src/blind150/twoPointers/ValidPalindrome.py
from curses.ascii import isalpha, isdigit

from src.utils.logging_util import get_logger

logger = get_logger(__name__)

class ValidPalindrome:
    def isPalindrome(self, s: str) -> bool:
        start, end = 0, len(s) - 1

        while start < end:
            while start < end and not s[start].isalpha() and not s[start].isdigit():
                logger.debug(f"Skipping non-alphanumeric character at start: {s[start]}")
                start += 1

            while start < end and not s[end].isalpha() and not s[end].isdigit():
                logger.debug(f"Skipping non-alphanumeric character at end: {s[end]}")
                end -= 1

            if s[start].lower() != s[end].lower():
                logger.warn(f"Mismatch found: {s[start]} != {s[end]} at positions {start} and {end}")
                return False

            start += 1
            end -= 1

        logger.info("The string is a palindrome.")
        return True


# Driver code
if __name__ == "__main__":
    valid_palindrome = ValidPalindrome()

    test_cases = [
        "A man, a plan, a canal: Panama",
        "race a car",
        " ",
        "No 'x' in Nixon",
        "hello",
        "madam"
    ]

    for test_case in test_cases:
        logger.info(f"Checking if '{test_case}' is a palindrome...")
        result = valid_palindrome.isPalindrome(test_case)
        logger.info(f"Is '{test_case}' a palindrome? {result}")

File: java/src/main/java/dsa/blind150/twoPointers/ValidPalindrome.java
package dsa.blind150.twoPointers;

import dsa.utils.LoggerUtil;
import org.slf4j.Logger;

public class ValidPalindrome {
    private static final Logger logger = LoggerUtil.getLogger();

    public boolean isPalindrome(String s) {
        int start = 0, end = s.length() - 1;
        while (start < end) {
            while (start < end && !Character.isLetterOrDigit(s.charAt(start))) {
                logger.debug("Skipping non-alphanumeric character at start: {}", s.charAt(start));
                start++;
            }
            while (start < end && !Character.isLetterOrDigit(s.charAt(end))) {
                logger.debug("Skipping non-alphanumeric character at end: {}", s.charAt(end));
                end--;
            }
            if (Character.toLowerCase(s.charAt(start)) != Character.toLowerCase(s.charAt(end))) {
                logger.warn("Mismatch found: {} != {} at positions {} and {}",
                        s.charAt(start), s.charAt(end), start, end);
                return false;
            }
            start++;
            end--;
        }
        logger.info("The string '{}' is a palindrome", s);
        return true;
    }

    public static void main(String[] args) {
        ValidPalindrome validPalindrome = new ValidPalindrome();

        String[] testCases = {
                "A man, a plan, a canal: Panama", // Expected: true
                "race a car",                    // Expected: false
                " ",                             // Expected: true
                "No 'x' in Nixon"                // Expected: true
        };

        for (String testCase : testCases) {
            logger.info("Checking if '{}' is a palindrome", testCase);
            boolean result = validPalindrome.isPalindrome(testCase);
            logger.info("Result: {}", result ? "Palindrome" : "Not Palindrome");
        }
    }
}